add errorsink.d, an abstract interface to error messages

WalterBright opened the pull request:
I've made considerable progress into making the lexer standalone, and the lexer/parser standalone. The problem is its reliance on globals.d and errors.d, which pull in a bunch of other stuff not relevant to lexing/parsing. The code in errors.d is of rather stupefying complexity, which is completely irrelevant to the lexer/parser. This complexity also makes it difficult to create a standalone lexer/parser.

Enter the abstract interface for error messages, ErrorSink, in errorsink.d. It's only got the functions in it that the lexer/parser actually need. The lexer/parser should not need to know anything about color formatting, highlighting, gagging, or where the messages are sent to.

The constructors for the lexer/parser are passed an instance of ErrorSink.

This isn't perfect, there's still an ugly wart in it that relies on getenv. I'll tackle that problem after this one is incorporated.

Sorry about the high line count, although nearly all of it is just changing the call to `error()`. The substantive changes are pretty small.

The more we can reduce all these "everybody imports everybody else" the more tractable the code base will be.

WalterBright committed:
add errorsink.d, an abstract interface to error messages

ibuclaw commented:
Errrrrrrrrrrrr @jacob-carlborg ?

ibuclaw commented:
#9350
#9484
#9480
#9507
#10072
#10711

jacob-carlborg commented:
🤦

RazvanN7 commented on compiler/src/dmd/lexer.d:
```diff
@@ -23,6 +23,7 @@ import core.stdc.time;
 
 import dmd.entity;
 import dmd.errors;
```
Shouldn't the use of errorsink get rid of this import?

WalterBright replied:
Yes, it should. But there are a couple other cases where it is referenced directly that I will fix in a subsequent PR.

thewilsonator commented:
Yay, lets add even more files before packaging dmd properly.

ibuclaw commented:
> Yay, lets add even more files before packaging dmd properly.

Worse than that, this PR literally [reinvents the wheel](https://github.com/dlang/dmd/blob/aa19d8a145afeb627a1136bf58f958998cd0fcec/compiler/src/dmd/frontend.d#L471-L683)... why?

thewilsonator commented on compiler/src/dmd/errorsink.d:
```diff
@@ -0,0 +1,121 @@
+/***************************************
+ * Where error/warning/deprecation messages go.
+ */
+abstract class ErrorSink
```
Reuse [existing wheels](https://github.com/dlang/dmd/blob/aa19d8a145afeb627a1136bf58f958998cd0fcec/compiler/src/dmd/frontend.d#L471-L683) or explain why they are insufficient. 

Such deficiencies might include `dmd.frontend` imports `std`.

Note also that `dmd.frontend` is a stable externally available API and so at worst this PR should move the definition from there to here, and have a publicly imported alias to the new place.

WalterBright replied:
> Reuse existing wheels or explain why they are insufficient.

This line illustrates the insufficiency:
```
Returns: false if the message should also be printed to stderr, true otherwise
```
This is the wrong abstraction. The user of Lexer should be able to send the errors anywhere, baking into the API "stderr" or not is insufficent. `ErrorSink` puts no constraint on where the messages should go. They can even go nowhere, as `ErrorSinkNull` demonstrates.

It's other insufficiency is it is centered around a pointer to a function which overrides the default behavior. This is a primitive form of OOP that fell out of favor in the 1980s. The abstract class, for which the user adds derived classes with specific behaviors, is the more modern approach. The abstract class designer does not have to anticipate what the implementation classes will be doing.

WalterBright replied:
I don't want to touch dmd.frontend in this PR. As I mentioned to @ibuclaw it needs a significant overhaul.

thewilsonator replied:
Then I consider this PR to require approval from @jacob-carlborg  before merging.

> don't want to touch dmd.frontend in this PR

I do not trust you will ever get around to this. And this adds yet more files without packaging things properly and I trust even less that you will ever do that.

> They can even go nowhere, as ErrorSinkNull demonstrates.

So can `DiagnositReporter`, `return true;`.

> As I mentioned to @ibuclaw it needs a significant overhaul.

Where? why? (Hint, no it doesn't).

ljmf00 reviewed (dismissed):
Please explain why we defer the creation of a proper error handling mechanism on DMD and also why not use the available interfaces exposed to be used on diagnostics?

WalterBright commented:
I'm mystified by the unittest runner:
```
./build unittest
(DC) LEXER-unittest
(DC) DMD-unittest
(RUN) DMD-UNITTEST
Success
```
However, the unittests were not run, as I put a printf in them. So looking at the -v output, I see it is linking in:
```
../generated/linux/release/64/lexer-unittest.o ../generated/linux/release/64/backend-unittest.o ../generated/linux/release/64/common-unittest.o
```
What? The unittest feature in D is designed so you compile with -unittest, run the program, and the unittests are run. Where are all these extra .o files for? And why won't the unittests run when I run the generated dmd? What happened to our simple way of doing unittests?

WalterBright commented:
> Why?

The best way to make understandable code is to have all the inputs and outputs at the entry point to the function. John Carmack explains this far better than I in http://www.sevangelatos.com/john-carmack-on/

What I've been doing over several previous PRs is just that - giving Lexer everything it needs via the call to its constructor, rather than "side loading" things through global variables, like `global`. This PR addresses a remaining problem - where do error messages go? The Lexer should not care what happens to the error messages, it just needs a place to send them. This is abstracted into ErrorSink, which is just what it says, it's a sink for error messages. Taking a look at the abstract class, it's utterly trivial and can be understood completely in 30 seconds. A couple handy implementations of it are there as well - ErrorSinkNull which just does nothing, and ErrorSinkStderr which just writes the messages to stderr. Both are trivial.

Contrast this with dmd.errors. It's a bewildering panoply of functions. There's a hook in it, `diagnosticHandler`, which snatches away control. But the user still has to understand all that stuff in errors.d, and it still all gets pulled into the executable, along with all its dependencies. It's full of global state, exactly what is undesirable in encapsulating functionality.

This PR does not break diagnosticHandler, or change the existing code that interfaces with it, other than adding a parameter to the constructor of Lexer.

Everything works but the unittests. So far, I have not been able to debug that because I have not managed to make heads or tails of how the unittests are run. Something about unittests in dmd have gone completely off the rails into a swamp.

WalterBright commented:
The following lines are in lexer.d:
```
if (!errorSink) { debug fprintf(stderr, "file %s\n", file.ptr); }
if (!errorSink) { debug printf("file %s\n", file.ptr); }
        assert(errorSink);
```
and the log file result is:
```
1) test CRLF and mixed line endings

core.exception.AssertError@src\dmd\lexer.d(133): Assertion failure
```
Why does neither printf show up in the log? This makes things very difficult to debug.

WalterBright commented:
> Worse than that, this PR literally reinvents the wheel ... why?

dmd.frontend provides an implementation, not an abstract interface. An abstract interface is much more adaptable. dmd.errorsink is an example of an abstract interface.

You're right that all those imports of `std`, like `std.algorithm`, are a deficiency. The compiler should be free of dependency on Phobos working.

dmd.frontend needs a modern redesign, but that is way beyond the scope of this PR for the time being.

Oddly enough, I am listed as the author of frontend.d. I didn't write it.

WalterBright commented on compiler/src/dmd/lexer.d:
```diff
@@ -2647,28 +2655,34 @@ class Lexer
     }
 
-    final void error(const(char)* format, ...)
+    final void error(const(char)* format)
```
They're just convenience functions. I suppose they could be marked as protected, but not in this PR.

jacob-carlborg commented:
> Then I consider this PR to require approval from @jacob-carlborg before merging.

I'm not doing any code reviews. This PR is the perfect example of why I'm not contributing anymore.

Geod24 commented on compiler/src/dmd/dmodule.d:
```diff
@@ -759,14 +760,17 @@ extern (C++) final class Module : Package
         Package ppack = null;
 
+        if (!global.errorSink)
+            global.errorSink = new ErrorSinkCompiler;
```
Why is global assigned to here ?

Geod24 commented on compiler/src/dmd/dmodule.d:
```diff
@@ -1382,7 +1386,7 @@ extern (C++) struct ModuleDeclaration
  * Params:
  *      mod = the Module
- *	aclasses = array to fill in
+ *      aclasses = array to fill in
```
Unrelated change, should be its own commit

WalterBright replied:
Another case of tabs being converted to spaces by my checkin procedure. Not worth another PR.

Geod24 commented on compiler/src/dmd/doc.d:
```diff
@@ -35,6 +35,7 @@ import dmd.dsymbol;
 import dmd.dsymbolsem;
 import dmd.dtemplate;
 import dmd.errors;
+import dmd.errorsink;
```
Why do we add this import if we're just passing a variable without ever referencing its type ?

WalterBright replied:
removed

Geod24 commented on compiler/src/dmd/dsymbolsem.d:
```diff
@@ -39,6 +39,7 @@ import dmd.dsymbol;
 import dmd.dtemplate;
 import dmd.dversion;
 import dmd.errors;
+import dmd.errorsink;
```
Ditto import

Geod24 commented on compiler/src/dmd/dtemplate.d:
```diff
@@ -7561,7 +7561,7 @@ extern (C++) class TemplateInstance : ScopeDsymbol
                 if (!s.isStaticIfDeclaration())
                 {
-		    //s.dsymbolSemantic(sc2);
+                    //s.dsymbolSemantic(sc2);
```
Unrelated, own commit

WalterBright replied:
That came about because of my checkin process converts all tabs to spaces, and apparently there were tabs on that line. Being a bit of trivia, there's no issue here.

Geod24 commented on compiler/src/dmd/errors.d:
```diff
@@ -24,6 +25,57 @@ import dmd.console;
+/***************************
+ * Error message sink for D compiler.
+ */
+class ErrorSinkCompiler : ErrorSink
```
I don't think that's the right abstraction either. What about messages issued by `pragma(msg)` ?

maxhaton replied:
```
bool pragmaMsgSemantic(Loc loc, Scope* sc, Expressions* args)
{
    if (!args)
        return true;
    foreach (arg; *args)
    {
        sc = sc.startCTFE();
        auto e = arg.expressionSemantic(sc);
        e = resolveProperties(sc, e);
        sc = sc.endCTFE();

        // pragma(msg) is allowed to contain types as well as expressions
        e = ctfeInterpretForPragmaMsg(e);
        if (e.op == EXP.error)
        {
            errorSupplemental(loc, "while evaluating `pragma(msg, %s)`", arg.toChars());
            return false;
        }
        if (auto se = e.toStringExp())
        {
            const slice = se.toUTF8(sc).peekString();
            fprintf(stderr, "%.*s", cast(int)slice.length, slice.ptr);
        }
        else
            fprintf(stderr, "%s", e.toChars());
    }
    fprintf(stderr, "\n");
    return true;
}
```
impl should really be using `message`, but anyway motivates the need for another category of output in this type of interface.

Geod24 commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -41,6 +41,7 @@ import dmd.dstruct;
 import dmd.dsymbolsem;
 import dmd.dtemplate;
 import dmd.errors;
+import dmd.errorsink;
```
Useless import

WalterBright replied:
yes, leftover from a previous incarnation of this PR

Geod24 commented on compiler/src/dmd/globals.d:
```diff
@@ -346,6 +350,8 @@ extern (C++) struct Global
 
     extern (C++) void _init()
     {
+        global.errorSink = new ErrorSinkCompiler;
```
This should be moved out of `globals` and into main to reduce circular dependencies

Geod24 commented on compiler/src/dmd/iasmgcc.d:
```diff
@@ -384,6 +385,9 @@ unittest
     import dmd.mtype : TypeBasic;
 
+    if (!global.errorSink)
+        global.errorSink = new ErrorSinkCompiler;
```
Likewise, this pattern is not good

WalterBright replied:
You're right, that's left over from a previous incarnation of this PR. I had a fair amount of trouble figuring out where the unittests were being initialized.

WalterBright replied:
Unfortunately, this code doesn't call globals._init, and so global.errorSink is not initialized. So I have to initialize it here.

ibuclaw replied:
Having everything in global is not ideal either.

Geod24 commented on compiler/src/dmd/mars.d:
```diff
@@ -40,6 +40,7 @@ import dmd.dsymbolsem;
 import dmd.dtemplate;
 import dmd.dtoh;
 import dmd.errors;
+import dmd.errorsink;
```
Useless import

WalterBright replied:
Yes, that's leftover from a previous incarnation of this PR.

Geod24 commented on compiler/test/unit/lexer/location_offset.d:
```diff
@@ -16,7 +17,7 @@ unittest
     enum code = "token";
 
-    scope lexer = new Lexer("test.d", code.ptr, 0, code.length, 0, 0);
+    scope lexer = new Lexer("test.d", code.ptr, 0, code.length, 0, 0, new ErrorSinkStderr);
```
I think this should just be a function at this point, instead of copy-pasted code

WalterBright replied:
It's a class because it consists of several functions, not one.

Geod24 replied:
I meant that at some point, we should realize that:
```D
scope lexer = new TestLexer(code);
// Or:
scope lexer = makeTestLexer(code);
```

is the way to go.

WalterBright replied:
That's worth considering for a future PR. We already use both techniques in dmd.

Geod24 reviewed (dismissed):
- Lots of useless imports;
- The lazy instantiation of globals in the middle of semantic is a big no;
- The two classes you provide, `ErrorSinkNull` and `ErrorSinkStderr`, are completely unused, something you have repeatedly complained about;

Instead of re-doing what other people have been wanting for years, perhaps we should just look into how/what @jacob-carlborg did, with `DiagnosticReporter`, which was also introduced gradually (starting with the lexer).

WalterBright commented:
> Lots of useless imports;

Removed. Explained in the comments.

> The lazy instantiation of globals in the middle of semantic is a big no;

Removed. Explained in the comments.

> The two classes you provide, ErrorSinkNull and ErrorSinkStderr, are completely unused, something you have repeatedly complained about;

ErrorSinkNull I plan to use in errors.d to replace the use of:
```
auto gaggedErrorsSave = global.startGagging();
```
in errors.d, which is the perfect use case for it.

I plan to use ErrorSinkError in the unit tests for Lexer.

Both are also an illustration for anyone needing their own custom error message sink.

> Instead of re-doing what other people have been wanting for years, perhaps we should just look into how/what @jacob-carlborg did, with DiagnosticReporter, which was also introduced gradually (starting with the lexer).

ErrorSink is a better design. It's hard to think of anything simpler - if you can, I want to hear about it. Note also that the only import ErrorSink needs is dmd.location. I'd actually like to get it to zero imports. In general, using sinks is an excellent approach to eliminating arbitrary dependencies.

WalterBright commented:
This serves our goals of reducing technical debt and making the lexer/parser more amenable to being standalone by reducing its dependencies on other modules. It's ready to merge. @RazvanN7 @dkorpel

dkorpel reviewed (dismissed):
Creating a copy of what already exists does not reduce technical debt. If there's deficiencies in `DiagnosticReporter`, why not fix that instead?

WalterBright commented:
> Creating a copy of what already exists does not reduce technical debt.

To reiterate:
1. it results in lexer.d having a dependency on dmd.console.
2. if a diagnosticreporter is used, the code for the non-use case remains in the program, unused
3. I quote: 
```
 * The diagnostic handler.
 * If non-null it will be called for every diagnostic message issued by the compiler.
 * If it returns false, the message will be printed to stderr as usual.
```
it should not leave a builtin dependency on stderr. It should not have default behavior at all.
4. It uses an old-fashioned pointer-to-handler OOP design that was supplanted in 1990 (aka a "hook" interface)
5. abstract interfaces are a better design
6. It still leaves a dependency on the gagging global which is an execrable design practice
7. It is much more complicated than necessary
8. anyone can understand ErrorSink in 30 seconds
9. GUI programs don't have a console or stderr
10. it enables easy mock sinks, which are critical for unittesting
11. it "side loads" into Lexer, rather than going through the function parameter list

This PR does not delete diagnostichandler because that would break existing code. I wished to keep this PR as a refactoring. Migrating from diagnostichandler to the abstract interface can be done later.

WalterBright commented on compiler/src/dmd/lexer.d:
```diff
@@ -539,7 +548,7 @@ class Lexer
                             if (isUniAlpha(u))
                                 continue;
-                            error(t.loc, "char 0x%04x not allowed in identifier", u);
+                            eSink.error(t.loc, "char 0x%04x not allowed in identifier", u);
```
It's problematic because the variadic argument list is clumsy to just forward things along.

dkorpel replied:
It's not:
```
    final void error(T...)(const(char)* format, T args)
    {
        error(token.loc, format, args);
    }

    final void error(T...)(const ref Loc loc, const(char)* format, T args)
    {
        error(loc, format, args);
    }
```
What's really clumsy is an awkward mix of calls to `error` and `eSink.error` based on the number of arguments.

WalterBright replied:
You're right. I didn't think of the template solution.

maxhaton commented on compiler/src/dmd/errorsink.d:
```diff
@@ -0,0 +1,121 @@
+
+    void errorSupplemental(const ref Loc loc, const(char)* format, ...);
+
+    void warning(const ref Loc loc, const(char)* format, ...);
```
`warningSupplemental`?

WalterBright replied:
I only added the ones used.

maxhaton commented on compiler/src/dmd/errorsink.d:
```diff
@@ -0,0 +1,121 @@
+    void deprecation(const ref Loc loc, const(char)* format, ...);
+
+    void deprecationSupplemental(const ref Loc loc, const(char)* format, ...);
+}
```
`message` should also be included.

WalterBright replied:
I only added the ones used. Others can be added as necessary.

maxhaton commented:
Is this any less dependant on gagging?

WalterBright commented:
> Is this any less dependant on gagging?

Currently, no. But as I remarked upthread, there's an obvious case in colorHightlightCode() for using ErrorSinkNull instead of gagging. Eventually, I'd like to see all use of gagging replaced by ErrorSinkNull.

ibuclaw commented on compiler/src/dmd/iasmgcc.d:
```diff
@@ -72,7 +73,7 @@ int parseExtAsmOperands(Parser)(Parser p, GccAsmStatement s)
                 else
                 {
-                    p.error(s.loc, "expected identifier after `[`");
+                    p.eSink.error(s.loc, "expected identifier after `[`");
```
This just makes iasmgcc ugly to look at for no obvious benefit.

ibuclaw commented on compiler/src/dmd/iasmgcc.d:
```diff
@@ -406,7 +410,7 @@ unittest
         const errors = global.errors;
         scope gas = new GccAsmStatement(Loc.initial, tokens);
-        scope p = new Parser!ASTCodegen(null, ";", false);
+        scope p = new Parser!ASTCodegen(null, ";", false, global.errorSink);
```
At this point, might as well make it a default param instead, there are dozens of repetitions of this pattern of change.

ibuclaw commented on compiler/src/dmd/errors.d:
```diff
@@ -24,6 +25,57 @@ import dmd.console;
+  extern (C++):
+  override:
+
+    void error(const ref Loc loc, const(char)* format, ...)
+    {
+        va_list ap;
+        va_start(ap, format);
+        verror(loc, format, ap);
+        va_end(ap);
+    }
```
Dangerously close to hijacking gcc error diagnostics here. Or am I expected to just fork out the implementation to the C++ side of the compiler?

ntrel commented on compiler/src/dmd/errorsink.d:
```diff
@@ -0,0 +1,121 @@
+
+    void error(const ref Loc loc, const(char)* format, ...)
+    {
+        fputs("Error: ", stderr);
```
Shouldn't `loc` come before "Error:" etc?
